### PR TITLE
Gsicard poc

### PR DESCRIPTION
**Pull request by gsicard**
First code review

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,162 @@
...
+
+**Important note:** if a syrah `File` is opened in writing mode outside a context manager (`with` statement) it needs to be closed explicitly using the `File.close()` method to ensure that the metadata is written at the end of the file and that the headers are properly updated.
+ 
+## 3. Reading from a syrah file
+
+Similarly to writing, the whole sample can be read at once and returns a dictionary of arrays:
+
+```python
+with File(file_path, mode='r') as syr:
+    for i in range(num_samples):
```
Comment (ilcord) [inline README.md]: specify that num_samples can be provided by len(syr)

File: syrah/__init__.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    This format is designed to store indexed dictionaries of Numpy arrays (called items) on disk
+    while allowing for fast random access of the items from their index using an in-memory index table.
+    It supports both fixed length and variable length array storage.
+
+    The data is stored in a binary file as follows:
+    <FORMAT VERSION><METADATA OFFSET><METADATA LENGTH><MAGIC BYTES><BYTE ARRAYS><SERIALIZED METADATA>
+"""
+
+# TODO: add import error checking
+# TODO: add setup.py
```
Comment (ilcord) [inline syrah/__init__.py]: already added

File: syrah/_hl/files.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+        self._mode = mode
+
+        if self._fp is not None:
+            self.close()
+
+        self._fp = open(self._file_path, self._mode + 'b')
+
+    def _init_data(self):
+        """
+        Initialize metadata anf file headers.
```
Comment (ilcord) [inline syrah/_hl/files.py]: *and* file headers

File: syrah/_hl/files.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+        :return:
+        """
+        if self._mode == 'w':
```
Comment (ilcord) [inline syrah/_hl/files.py:71]: add separate methods for read and write.

File: syrah/_hl/files.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+            self._version = version.version
+            self._metadata_offset = 0
+            self._metadata_length = 0
+            self._write_headers()
```
Comment (ilcord) [inline syrah/_hl/files.py]: why writing headers before metadata length and offset were initialized?

File: syrah/_hl/files.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+        if self._mode == 'w':
+            self._flush()
+
+        self._fp.close()
+
+    def get_item(self, item: int) -> Dict[str, ndarray]:
+        """
+        Get an item from the dataset.
+        :param item: index of the item in the dataset
+        :return: a dictionary of arrays
```
Comment (ilcord) [inline syrah/_hl/files.py]: didn't specify what's the key of the returned dictionary

File: syrah/_hl/files.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+        """
+        if self._fp is None:
```
Comment (ilcord) [inline syrah/_hl/files.py]: refactor to validate_input_file()

File: syrah/_hl/files.py [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+            self._fp.seek(self._metadata.get(item, key, 'offset'))
+            array_serialized: AnyStr = self._fp.read(self._metadata.get(item, key, 'size'))
+            data[key]: ndarray = deserialize_array(array_serialized, self._metadata.get(item, key, 'dtype'))
+
+        return data
+
+    def get_array(self, item: int, key: str) -> ndarray:
+        """
+        Get an array from the dataset.
+        :param item: index of the item in the dataset
```
Comment (ilcord) [inline syrah/_hl/files.py]: rename item to index

File: syrah/_hl/files.py [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+        :param key: key of the array to retrieve
+        :return: an array
+        """
+        if self._fp is None:
```
Comment (ilcord) [inline syrah/_hl/files.py]: move to intput validation method

File: syrah/_hl/files.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+        :return: number of items
+        """
+        return len(self._metadata)
+
+    def _read_headers(self):
+        """
+        Read the headers and the metadata from file.
+        :return:
+        """
+        if self._fp.closed:
```
Comment (ilcord) [inline syrah/_hl/files.py]: move to input validation method

File: syrah/_hl/files.py [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+        self._version = self._version_to_string(header_version)
+        self._metadata_offset = np.frombuffer(header_metadata_offset, dtype=np.int64)[0]
+        self._metadata_length = np.frombuffer(header_metadata_length, dtype=np.int64)[0]
+
+    def _write_headers(self):
+        """
+        Write headers to file.
+        :return:
+        """
+        if self._fp.closed:
```
Comment (ilcord) [inline syrah/_hl/files.py]: move to output validation method

File: syrah/_hl/files.py [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+        self._fp.seek(0)
+        self._fp.write(config.MAGIC_BYTES)
+        self._fp.write(header_version)
+        self._fp.write(header_metadata_offset)
+        self._fp.write(header_metadata_length)
+
+    def add_item(self, data: Dict[str, ndarray]):
+        """
+        Write an item with the given index to the dataset.
+        :param item: index of the item
```
Comment (ilcord) [inline syrah/_hl/files.py]: there's no item in the actual code

File: syrah/_hl/files.py [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+        :param data: dictionary of arrays
```
Comment (ilcord) [inline syrah/_hl/files.py]: what does the str key stands for?

File: syrah/_hl/files.py [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+        :return:
+        """
+        if self._fp is None:
```
Comment (ilcord) [inline syrah/_hl/files.py]: move to validation method

File: syrah/_hl/files.py [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+                raise ValueError(f'Expected array_key type to be str, got {type(array_key)}.')
+            if type(array_value) is not ndarray:
+                raise ValueError(f'Expected value type to be ndarray, got {type(array_value)}.')
+
+            array_serialized, dtype = serialize_array(array_value)
+
+            self._fp.seek(item_offset)
+            self._fp.write(array_serialized)
+
+            array_metadata = dict()
```
Comment (ilcord) [inline syrah/_hl/files.py]: refactor to create_array_metadata()

File: syrah/_hl/files.py [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+            item_offset += len(array_serialized)
+
+        self._metadata.add_item(metadata_item)
+        self._item_offset = item_offset
+
+    @staticmethod
+    def _version_to_string(version_bytes: bytes) -> str:
+        """
+        Deserialize version number from bytes to string format.
+        :param version_bytes: 4-byte array representation of the version number
```
Comment (ilcord) [inline syrah/_hl/files.py:277]: specify that the binary was encoded in  *little endian*  for all encoding/decoding methods

File: syrah/_hl/files.py [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+        :return: string version with format "x.y.z"
+        """
+        return '.'.join([str(x) for x in np.frombuffer(version_bytes, dtype=np.uint8)[1:]])
+
+    @staticmethod
+    def _version_to_bytes(version_string: str) -> bytes:
+        """
+        Serialize version number from string format.
```
Comment (ilcord) [inline syrah/_hl/files.py]: specify little endian

File: syrah/_hl/files.py [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+            raise IOError(f'File is expected to be opened in write mode, got {self._mode}.')
+
+        metadata_serialized: AnyStr = self._metadata.tobytes()
+
+        self._metadata_offset = self._item_offset
+        self._metadata_length = len(metadata_serialized)
+        self._write_headers()
+
+        self._fp.seek(self._item_offset)
+        self._fp.write(metadata_serialized)
```
Comment (ilcord) [inline syrah/_hl/files.py:308]: add close() in the end of flush() to eliminate undefined behavior

File: syrah/_hl/metadata.py [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+metadata_ctypes = {
+    'offset': [ctypes.c_int64],
+    'size': [ctypes.c_int64],
+    'dtype': str
+}
+
+# TODO: wrapp arrays with multiprocessing.Array: https://stackoverflow.com/questions/5549190/is-shared-readonly-data-copied-to-different-processes-for-multiprocessing/5550156#5550156
+class AbstractMetadata:
```
Comment (ilcord) [inline syrah/_hl/metadata.py]: No need for the abstract class. Reader and writer are not really different implementations of metadata, they actually implement opposite functionalities. If you think that we'll have different reader/writer implementations, I would split this class to AbstractMetadataReader/Writer classes with concrete implementations. However, for the lack of plausibility in that direction, I would discard this class completely

File: syrah/_hl/metadata.py [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+                                         f' {num_items} != {length}')
+
+        self.length = length
+
+
+class MetadataWriter(AbstractMetadata):
+    """
+    Represent a metadata writer object.
+    """
+    def add_item(self, item_metadata: Dict[AnyStr, Dict[AnyStr, int]]):
```
Comment (ilcord) [inline syrah/_hl/metadata.py:87]: int -> np.int64

File: syrah/_hl/metadata.py [hunk 21]
```diff
@@ -0,0 +1,187 @@
...
+        """
+        Add a new item to the metadata.
+        :param item_metadata: dictionary of array metadata
+        :return:
+        """
+        # TODO: check metadata types consistency with expected type
```
Comment (ilcord) [inline syrah/_hl/metadata.py]: method too long, split to smaller ones

File: syrah/_hl/metadata.py [hunk 22]
```diff
@@ -0,0 +1,187 @@
...
+            raise KeyError('Array keys not consistent with previous metadata.')
+
+        for array_key, array_metadata in item_metadata.items():
+            if set(array_metadata.keys()) != set(metadata_types.keys()):
+                raise KeyError(f'Metadata keys: {", ".join(array_metadata.keys())} do not match requirements:'
+                               f' {", ".join(metadata_types.keys())}.')
+
+            for metadata_key in metadata_types.keys():
+                if isinstance(metadata_types[metadata_key], list):
+                    self.metadata[array_key][metadata_key] += np.append(self.metadata[array_key][metadata_key],
```
Comment (ilcord) [inline syrah/_hl/metadata.py]: doesn't make sense - you've already added the key to the metadata, and then you're appending the result to the metadata?

File: syrah/_hl/serialization.py [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+    19: 'str'
+}
+
+"""
+Supported types by name
+"""
+dtype_names = {value: key for key, value in dtypes.items()}
+
+
+def format_dtype(data_type: dtype) -> str:
```
Comment (ilcord) [inline syrah/_hl/serialization.py]: to_string() ?

File: syrah/_hl/serialization.py [hunk 24]
```diff
@@ -0,0 +1,100 @@
...
+    if dtype_num not in dtypes:
+        raise TypeError(f'Type {data_type} is not supported. Supported types are: {", ".join(dtype_names.keys())}')
+
+    return dtypes[dtype_num]
+
+
+def deserialize_array(array_serialized: AnyStr, data_type: str) -> ndarray:
+    """
+    Deserialize an array using the given type.
+    :param array_serialized: serialized array as byte string
```
Comment (ilcord) [inline syrah/_hl/serialization.py:73]: specify little endian

File: syrah/_hl/serialization.py [hunk 25]
```diff
@@ -0,0 +1,100 @@
...
+    :param data_type: type of the resulting array
+    :return: numpy array of the specified type
+    """
+    if data_type not in dtype_names:
+        raise TypeError(f'Type {data_type} is not supported. Supported types are: {", ".join(dtype_names.keys())}')
+
+    if data_type == 'str':
```
Comment (ilcord) [inline syrah/_hl/serialization.py:80]: specify str special case in the documentation

File: syrah/_hl/serialization.py [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+
+def serialize_array(array: ndarray) -> Tuple[AnyStr, str]:
+    """
+    Serialize a numpy array to byte string
+    :param array: numpy array
+    :return: serialized array and corresponding array type
+    """
+    data_type = format_dtype(array.dtype)
+
+    if data_type == 'str':
```
Comment (ilcord) [inline syrah/_hl/serialization.py:97]: specify string special case in documentation

File: syrah/utils/data.py [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+"""
+from typing import List, Tuple
+from torch.utils.data.dataset import Dataset, ConcatDataset
+from numpy import ndarray
+from .. import File
+
+
+class SyrahDataset(Dataset):
+    """
+    Represent a PyTorch Dataset using a Syrah file to store the data.
```
Comment (ilcord) [inline syrah/utils/data.py]: Represent*s*

File: syrah/utils/data.py [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+        """
+        Returns the length of the dataset.
+        :return: length of the dataset
+        """
+        return self.len
+
+
+class SyrahConcatDataset(ConcatDataset):
+    """
+    Represent a concatenation of syrah datasets.
```
Comment (ilcord) [inline syrah/utils/data.py]: Represents

File: syrah/utils/data.py [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+        :param file_path: path to the Syrah file
+        :param keys: list of keys to retrieve from the data
+        """
+        self.file_path = file_path
+        self.keys = keys
+
+        self.syr = File(file_path, 'r')
+        self.len = self.syr.num_items()
+
+    def open(self, worker_id: int = -1):
```
Comment (ilcord) [inline syrah/utils/data.py]: document worker_id param

File: syrah/version.py [hunk 30]
```diff
@@ -0,0 +1,25 @@
...
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Syrah.  If not, see <https://www.gnu.org/licenses/>.
+"""
+from collections import namedtuple
+
+_SYRAH_VERSION = namedtuple("_SYRAH_VERSION", "major minor bugfix")
```
Comment (ilcord) [inline syrah/version.py:21]: syrah version should be 2 bytes - major and minor

File: tests/test_syrah_file.py [hunk 31]
```diff
@@ -0,0 +1,144 @@
...
+
+class TestMultiMethods(SyrFileTestEnvironment, unittest.TestCase):
+    def test_sequential_item_read(self):
+        self.fp_syr = File()
+        item_idxs = range(self.num_items)
+
+        p = Pool(self.num_workers, initializer=self.fp_syr.open, initargs=(self.syr_path, 'r'))
+        p.map(self.assert_item_read, item_idxs)
+
+    # def test_sequential_array_read(self):
```
Comment (ilcord) [inline tests/test_syrah_file.py]: remove this

File: tests/test_syrah_file.py [hunk 32]
```diff
@@ -0,0 +1,144 @@
...
+if __name__ == '__main__':
+    test = TestSingleMethods()
+    test.setUp()
+    test.test_sequential_item_read()
+
+    test = TestMultiMethods()
+    test.setUp()
+    test.test_sequential_item_read()
+
+    # unittest.main()
```
Comment (ilcord) [inline tests/test_syrah_file.py]: remove this

File: README.md [hunk 33]
```diff
@@ -1,2 +1,162 @@
...
+Similarly, when using `num_workers > 0` with PyTorch `Dataloader`, `SyrahDataset.open()` also needs to be called by each worker (or `SyrahConcatDataset.open()` in case of multiple syrah files):
+
+```python
+data_generator_multi = DataLoader(dataset, batch_size=32, shuffle=True, num_workers=num_workers, worker_init_fn=dataset.open)
+
+for features, labels in data_generator_multi:
+    ...
+```
+
+## 6. File format summary
```
Comment (ilcord) [inline README.md:155]: specify that everything is in little endian

File: syrah/_hl/files.py (hunk 18 above)
Comment (gsicard) [reply syrah/_hl/files.py:308]: _flush is already called from close()

File: syrah/_hl/metadata.py (hunk 20 above)
Comment (gsicard) [reply syrah/_hl/metadata.py:87]: https://www.python.org/dev/peps/pep-0237/

File: syrah/version.py (hunk 30 above)
Comment (gsicard) [reply syrah/version.py:21]: Maybe in the file, not here

**Pull request merged**